bee-pow/src/providers/mod.rs

```diff
@@ -0,0 +1,40 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Contains nonce providers for Proof of Work.
+
+mod constant;
+mod miner;
+
+pub use constant::{Constant, ConstantBuilder};
+pub use miner::{Miner, MinerBuilder};
+
+use std::sync::{atomic::AtomicBool, Arc};
+
+/// A trait to build nonce providers.
+pub trait NonceProviderBuilder: Default + Sized {
+    /// The type of the built nonce provider.
+    type Provider: NonceProvider<Builder = Self>;
+
+    /// Creates a new nonce provider builder.
+    fn new() -> Self;
+
+    /// Constructs the nonce provider from the builder.
+    fn finish(self) -> Self::Provider;
```

Comment:
I propose that the `finish` method returns the `Arc<AtomicBool>` to cancel the miner. The `nonce` method signature then becomes simpler only taking `bytes` and `target_score`.

Reply:
But that would mean the Provider is only usable once, or we would have to reset the bool but it's a kinda weird API

Reply:
Hm, I don't think it would be a weird API tbh. The `Arc<AtomicBool` already says, that it's shared between the threads and the user, and anyone can change its value. So if the user wants to reuse the provider s/he can also expect that the `Arc<AtomicBool>` (which I maybe would wrap in a convenience type) has been reset.